sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanner.scala

```diff
@@ -44,8 +44,7 @@ class SparkPlanner(val sqlContext: SQLContext) extends SparkStrategies {
       EquiJoinSelection ::
       InMemoryScans ::
       BasicOperators ::
-      CartesianProduct ::
-      BroadcastNestedLoopJoin :: Nil)
+      CartesianProduct :: Nil)
```

Comment:
`BroadcastNestedLoopJoin` actually supports equi-join, thus we never run into this case, as we have more optimal solution for it in previous rules.

After a double think, I am a little hesitate to combine the rules in `CartesianProduct` and `BroadcastNestedLoopJoin`, as the later is supposed to be the last gate for JOIN, and works for all kinds of JOIN type w/ or w/o join condition, the others can be considered as the optimization compared to it.

I am going to revert the code change if @yhuai is not strongly opposite to it. Or we can refactor the JOIN strategy after this PR been merged.

What do you think @yhuai ?